src/software/embedded/robot_localizer.h

```diff
@@ -0,0 +1,144 @@
+#pragma once
+
+#include <Eigen/Dense>
+#include <deque>
+
+#include "proto/world.pb.h"
+#include "software/embedded/services/imu.h"
+#include "software/geom/angle.h"
+#include "software/geom/angular_velocity.h"
+#include "software/sensor_fusion/filter/kalman_filter.h"
+#include "time.h"
+
+/**
+ * This class implements the KalmanFilter class to track robot orientation. It keeps a
+ * rolling history of the robot's state, so when vision updates come in, the filter can
+ * roll back and recompute with them.
+ *
+ * For information on the math and design of kalman filters, see relevant links in
+ * kalman_filter.h
+ */
+class RobotLocalizer
+{
+   public:
+    RobotLocalizer(double process_noise_variance, double vision_noise_variance,
+                   double encoder_noise_variance,
+                   double target_angular_acceleration_variance)
+        : filter_(),
+          process_noise_variance_(
+              process_noise_variance),  // assumes process noise is a discrete time wiener
+                                        // process
+          history()
+    {
+        // Set state covariance, this is mostly a tuned value. The digaonal is the
+        // variance of the Orientation, Ang Vel, and Ang Accel
+        filter_.P << 30, 0, 0, 0, 4, 0, 0, 0, 5;
+        // Set control to state matrix, control space is a 1x1 matrix
+        filter_.B << 0.0, 1, 0.0;
+        // Set measurement variance.
+        filter_.R << vision_noise_variance, 0.0, 0.0, 0.0, 0.0, encoder_noise_variance,
+            0.0, 0.0, 0.0, 0.0, ImuService::IMU_VARIANCE, 0.0, 0.0, 0.0, 0.0,
+            target_angular_acceleration_variance;
+        filter_.x = Eigen::Matrix<double, 3, 1>(0.0, 0.0, 0.0);
+        clock_gettime(CLOCK_MONOTONIC, &current_time_);
+        clock_gettime(CLOCK_MONOTONIC, &last_step_);
+    }
+
+    void step(const AngularVelocity& targetAcceleration);
+
+    /** Update the orientation from vision based on an old reading.
+     *
+     * @param orientation Vision reading of the orientation of the robot in world space
+     * @param ageSeconds Age in seconds of the vision snapshot (time since it was taken)
+     */
+    void rollbackVision(const Angle& orientation, const double& ageSeconds);
```

Comment:
```suggestion
     * @param age_seconds Age in seconds of the vision snapshot (time since it was taken)
     */
    void rollbackVision(const Angle& orientation, const double& age_seconds);
```